### PR TITLE
Add backward implementation for LSTM operator. 

### PR DESCRIPTION
**Pull request by qingqing01**
Fix https://github.com/PaddlePaddle/Paddle/issues/5114

some enhancements will be done in next PR.
   - Now set the cell state grad to zero to keep the same with the old framework, will support to non-zero cell state gradient.
  - Support initial hidden state and cell state.
  - In this PR, use the fixed activations (`Sigmoid` and `Tanh`) since there is a bug for activation function pointer.  Will support to activations specified by users.

### CODE REVIEW

File: paddle/operators/lstm_op.h [hunk 1]
```diff
@@ -112,27 +123,198 @@ class LSTMKernel : public framework::OpKernel<T> {
...
+      // Following Eigen computation failed for double type on GPU device.
+      // bias_g->mutable_data<T>(ctx.GetPlace());
+      // Tensor bias_mat;
+      // bias_mat.ShareDataWith(*bias_g);
+      // bias_mat.Resize({1, 4 * frame_size});
+
+      // auto bias_g_e = EigenVector<T>::Flatten(bias_mat);
+      // auto gate_g_e = EigenMatrix<T>::From(batch_gate_g);
+      // Eigen::array<int, 1> dims{{0}};
+      // bias_g_e.device(ctx.GetEigenDevice<Place>()) = gate_g_e.sum(dims);
```
Comment (luotao1) [inline paddle/operators/lstm_op.h]: 295-304行是TODO？

File: paddle/operators/lstm_op.h [hunk 2]
```diff
@@ -112,27 +123,198 @@ class LSTMKernel : public framework::OpKernel<T> {
...
+      lstm_value.stateActiveValue = cell_pre_act.data<T>();
+
+      Tensor out_g = batch_hidden_g.Slice(bstart, bend);
+      Tensor gate_g = batch_gate_g.Slice(bstart, bend);
+      Tensor cell_g = batch_cell_g.Slice(bstart, bend);
+      lstm_grad.stateGrad = cell_g.data<T>();
+      lstm_grad.gateGrad = gate_g.data<T>();
+      lstm_grad.outputGrad = out_g.data<T>();
+
+      if (n != 0) {
```
Comment (luotao1) [inline paddle/operators/lstm_op.h]: `if (n != 0)` -》 `if (n)` ？

File: paddle/operators/math/math_function_test.cu [hunk 3]
```diff
@@ -177,3 +177,65 @@ TEST(math_function, gemm_trans_cublas) {
...
+  } else {
+    for (int i = 0; i < n; ++i) {
+      T sum = 0.0;
+      for (int j = 0; j < m; ++j) {
+        sum += data_a[j * n + i] * data_b[j];
+      }
+      ASSERT_FLOAT_EQ(data_c[i], sum);
+    }
+  }
+}
```
Comment (luotao1) [inline paddle/operators/math/math_function_test.cu]: 这里看上去和cc的单测，很多代码都是一样的。后续会考虑共用么？

**Comment by luotao1** [comment]
> some enhancements will be done in next PR.

具体是哪些呢？

File: paddle/operators/lstm_op.cc [hunk 4]
```diff
@@ -109,22 +123,27 @@ class LSTMOpMaker : public framework::OpProtoAndCheckerMaker {
              " - Bias = {b_c, b_i, b_f, b_o}."
              "2. `usePeepholes = True` "
              " - The shape is (1 x 7D). "
-             " - Bias = {b_c, b_i, b_f, b_o, W_ic, W_fc, W_oc}.");
+             " - Bias = {b_c, b_i, b_f, b_o, W_ic, W_fc, W_oc}.")
+        .AsDispensable();
+    AddOutput("Hidden",
+              "(LoDTensor) the hidden state lod tensor of LSTM operator. "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done. Remove lod tensor and fix the shape info.

File: paddle/operators/lstm_op.cc [hunk 5]
```diff
@@ -109,22 +123,27 @@ class LSTMOpMaker : public framework::OpProtoAndCheckerMaker {
...
+              "The shape and lod is the same with the `Input`.");
+    AddOutput("Cell",
+              "(LoDTensor) the cell state lod tensor of LSTM operator. "
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.cc [hunk 6]
```diff
@@ -109,22 +123,27 @@ class LSTMOpMaker : public framework::OpProtoAndCheckerMaker {
...
         .AsIntermediate();
-    AddOutput("Hidden",
-              "(LoDTensor) the hidden state lod tensor of LSTM operator. "
-              "The shape and lod is the same with the `Input`.");
-    AddOutput("Cell",
-              "(LoDTensor) the cell state lod tensor of LSTM operator. "
-              "The shape and lod is the same with the `Input`.");
+    AddOutput("BatchCellPreAct",
+              "(LoDTensor) This LoDTensor is get in the forward and used "
+              "in the backward.")
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.cc]: Done.

File: paddle/operators/lstm_op.h [hunk 7]
```diff
@@ -112,27 +123,198 @@ class LSTMKernel : public framework::OpKernel<T> {
...
+    auto* bias = ctx.Input<Tensor>("Bias");
+
+    auto* hidden_out = ctx.Input<LoDTensor>("Hidden");
+    auto* cell_out = ctx.Input<LoDTensor>("Cell");
+
+    auto* batch_gate = ctx.Input<LoDTensor>("BatchGate");
+    auto* batch_cell_pre_act = ctx.Input<LoDTensor>("BatchCellPreAct");
+
+    auto* hidden_g = ctx.Input<LoDTensor>(framework::GradVarName("Hidden"));
+    // auto* cell_g = ctx.Input<LoDTensor>(framework::GradVarName("Cell"));
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/lstm_op.h]: Done.

File: paddle/operators/lstm_op.h (hunk 2 above)
Comment (qingqing01) [reply paddle/operators/lstm_op.h]: Done.

File: paddle/operators/lstm_op.h (hunk 1 above)
Comment (qingqing01) [reply paddle/operators/lstm_op.h]: They are the equivalent code by Eigen, but the Eigen does not support double type on GPU device, so use GEMV.  And I remove these lines.

File: paddle/operators/math/math_function_test.cu (hunk 3 above)
Comment (qingqing01) [reply paddle/operators/math/math_function_test.cu]: 后续可以改进下。

Comment (wangkuiyi) [reply paddle/operators/math/math_function_test.cu]: 可以建立一个issue，把URL贴在这里就好。这样日后不会忘记了。

Comment (qingqing01) [reply paddle/operators/math/math_function_test.cu]: 建立个issue: https://github.com/PaddlePaddle/Paddle/issues/5234

**Pull request merged**